# Task 2 - Detecting Fraudulent Sales
You were just hired by BA Toys as an Internal Auditor. The Chief Audit Executive (CAE) is also new and has been reviewing processes surrounding the purchasing and sales processes. The internal controls appear to be well-designed, but the external auditor has raised some concerns about revenues. Specifically, the auditor claims that their analytics-based risk assessment suggests sales appear inconsistent with inventory levels. The CAE has reviewed the auditor’s evidence and agrees the issue warrants investigation.

The CAE has tasked you with reviewing inventory records, purchasing activity, and sales activity. She has obtained the data for you from the relevant departments. Below, she provides a description of each data source as well as what she can ascertain from the data. However, her information may be incomplete, and she conveys that the departments (Sales, Purchasing, and Inventory Management) are unlikely to be of much help. Like most retailers, BA’s fiscal year-end is January 31, and you will be focusing on the year ended January 31, 2018.


### Data files 📂
You will find the following datasets in your task materials, which you can unzip (if on ICE, I have provided the reference to these files):

1.	“BAToysEndInvJan2017.csv” and “BAToysEndInvJan2018.csv” – These files contain beginning and ending inventory, respectively, for BA Toys’ 21 stores.

2.	“BAToysPurchasesJan2018.csv” – All inventory purchases made during the year ending 1/31/2018.

3.	“BAToysTAXSalesJan2018.csv” – All sales occurring during the year ending 1/31/2018.

4.	“BAToysReturns.csv” – All returns processed during the year ending 1/31/2018. Note that returns refer to toys returned *to the vendor* (i.e., they reduce inventory available for sale). These *are not* customer returns.

5.	“PriceListing.csv” – A listing purchase and sales prices for inventory.


**Additional Notes** 📝

As noted, the CAE did not receive much information about the data, but made a few notes before passing the task to you:

- Each store is referenced in several datasets and appears to be referred to by a state code (e.g., “NY”) followed by a four-digit number.
- The CAE thinks BrandId somehow corresponds to InventoryId, which may give reference to stores.


### Requirements (summarized here, given in more detail below)
1. Review & describe data.
2. Inspect toy vendor products and comment on volume, margins, etc.
3. Determine whether any fraudulent sales appear to exist; quantify the dollar impact of fradulent sales.
4. (4046 only) Conduct qualitative review of suspected fraudulent sales.
5. (6046 only) Conduct systematic review of suspected fraudulent sales.


#### General Hint

The Sales, Purchases, and Returns files all have timestamps that would allow you to evaluate inventory at any given date. However, I suggest simplifying your analysis based on the specific question of whether there are fraudulent sales in the year (regardless of when they occur). In other words, begin by considering total purchases (by store-product), total sales (by store-product), etc. Collapse the datasets using the idenfier(s) corresponding to unique Store-Brand combinations. This may be sufficient to uncover any issues. If it is not, then you can circle back and perform a more detailed analysis at the daily level (but, hint, you shouldn’t have to).


## Requirement 1
Load each dataset into a pandas dataframe. Inspect the data and provide the CAE with definitions of what the various identifiers appear to mean.

Specifically, briefly describe each of the following fields:
1. Store
2. VendorNo (or VendorId)
3. Brand (or BrandId)
4. InventoryId

In [2]:
import pandas as pd

folder = "Task2Data" # update to the correct folder path

# Load datasets (I provided first, fill in the rest)

#For Desktop
beginv = pd.read_csv(f"{folder}/BAToysEndInvJan2017.csv")
endinv = pd.read_csv(f"{folder}/BAToysEndInvJan2018.csv")
purch  = pd.read_csv(f"{folder}/BAToysPurchasesJan2018.csv")
sale   = pd.read_csv(f"{folder}/BAToysTAXSalesJan2018.csv")
returns= pd.read_csv(f"{folder}/BAToysRetJan2018.csv")

listings = pd.read_csv(f"{folder}/PriceListing.csv") #Adding one more for references

# Add code here to inspect the datasets:

def printColumns (datasets: list, column: str):
    X = 1
    for data in datasets:
        print(f'Unique values for Dataset {X}: \n',data[column].unique())
        X += 1
print('For Store Columns:')
printColumns([beginv, endinv, purch, returns], 'Store')

#For printing Vendor ID
print('\nFor Vendor ID:')
print(sale.VendorNo.unique()) 
print(listings['VendorId'].unique())


For Store Columns:
Unique values for Dataset 1: 
 ['CT-6253' 'DC-3827' 'DE-2832' 'MA-1384' 'MA-1738' 'MA-2647' 'MA-5262'
 'MD-2830' 'MD-3274' 'ME-0293' 'NH-7575' 'NJ-4291' 'NJ-5264' 'NY-3349'
 'NY-3458' 'NY-7283' 'PA-1445' 'PA-7263' 'PA-9272' 'RI-4893' 'VT-8234']
Unique values for Dataset 2: 
 ['CT-6253' 'DC-3827' 'DE-2832' 'MA-1384' 'MA-1738' 'MA-2647' 'MA-5262'
 'MD-2830' 'MD-3274' 'ME-0293' 'NH-7575' 'NJ-4291' 'NJ-5264' 'NY-3349'
 'NY-3458' 'NY-7283' 'PA-1445' 'PA-7263' 'PA-9272' 'RI-4893' 'VT-8234']
Unique values for Dataset 3: 
 ['CT-6253' 'NJ-4291' 'MA-5262' 'VT-8234' 'MD-3274' 'MD-2830' 'ME-0293'
 'NH-7575' 'PA-7263' 'MA-2647' 'PA-9272' 'DE-2832' 'NJ-5264' 'DC-3827'
 'PA-1445' 'MA-1738' 'MA-1384' 'NY-3458' 'RI-4893' 'NY-7283' 'NY-3349']
Unique values for Dataset 4: 
 ['CT-6253' 'NJ-4291' 'MA-5262' 'MD-3274' 'NH-7575' 'MA-2647' 'MD-2830'
 'DE-2832' 'VT-8234' 'MA-1738' 'PA-9272' 'MA-1384' 'NY-7283' 'PA-7263'
 'ME-0293' 'NJ-5264' 'RI-4893' 'NY-3458' 'NY-3349' 'DC-3827' 'PA-1445']



In [3]:
#For printing Brand
print('\nFor Brand ID/No')
printColumns([beginv, endinv, purch, sale, returns], 'Brand')
print('Unique values for Dataset 6:\n', list(listings['BrandId'].sort_values(ascending = True))) 



For Brand ID/No
Unique values for Dataset 1: 
 [ 1001 10088 10108 ... 76364 29913  8353]
Unique values for Dataset 2: 
 [ 1001 10108 10169 ... 30673 43843 31749]
Unique values for Dataset 3: 
 [  964 17195 99340 ... 23675 43843 30203]
Unique values for Dataset 4: 
 [ 1001 10088 10108 ... 53604 14153 99654]
Unique values for Dataset 5: 
 [34471 36343  8070 ... 23675 61093 38256]
Unique values for Dataset 6:
 [49, 69, 108, 149, 152, 173, 178, 180, 222, 252, 276, 299, 335, 350, 363, 385, 400, 414, 457, 461, 489, 504, 558, 586, 648, 672, 679, 686, 764, 769, 770, 863, 921, 931, 964, 965, 982, 1001, 1012, 1018, 1031, 1032, 1040, 1081, 1090, 1101, 1105, 1135, 1152, 1163, 1164, 1173, 1185, 1189, 1194, 1199, 1202, 1203, 1232, 1234, 1240, 1256, 1262, 1267, 1274, 1296, 1306, 1332, 1335, 1338, 1448, 1455, 1480, 1541, 1562, 1578, 1604, 1634, 1648, 1652, 1683, 1689, 1711, 1757, 1801, 1805, 1815, 1849, 1866, 1867, 1889, 1914, 1920, 1939, 1947, 1965, 1967, 2036, 2067, 2072, 2089, 2091, 2106, 2113, 21

In [4]:
#For Printing VendorID
print('For Vendor ID')
printColumns([beginv, endinv, purch, sale, returns], 'InventoryId')

For Vendor ID
Unique values for Dataset 1: 
 ['CT-6253-1001' 'CT-6253-10088' 'CT-6253-10108' ... 'RI-4893-90645'
 'VT-8234-82202' 'VT-8234-88792']
Unique values for Dataset 2: 
 ['CT-6253-1001' 'CT-6253-10108' 'CT-6253-10169' ... 'VT-8234-73881'
 'VT-8234-74952' 'VT-8234-9428']
Unique values for Dataset 3: 
 ['CT-6253-964' 'CT-6253-17195' 'CT-6253-99340' ... 'NH-7575-75226'
 'NJ-5264-30203' 'ME-0293-97008']
Unique values for Dataset 4: 
 ['CT-6253-1001' 'CT-6253-10088' 'CT-6253-10108' ... 'VT-8234-99911'
 'VT-8234-9996' 'VT-8234-99981']
Unique values for Dataset 5: 
 ['CT-6253-34471' 'NJ-4291-36343' 'MA-5262-8070' ... 'NY-3349-48616'
 'NY-3349-24807' 'NY-3349-99410']


✅  *Insert Requirement 1 answer:*

1. <p style = 'color: #1ab9b9ff'>The Store identifier seems to note the specific store using a State identifier followed by a 4 digit code. If guessed, the first two letters are the state initials and the last 4 numbers appears to be a store identifier. However, not enough information is provided to determine the numbering system </p>

2.<p style = 'color: #1ab9b9ff'> Vendor ID/No seems to be a foreign key in the Sales Dataset for vendor information. The sales dataset includes information regarding the vendor name and vendor ID. It is highly probable that we use the vendor ID/No from the Sales Dataset to merge with the Price Listing. However, we note that a vendor can provide multiple products, so using only this identifier is not good enough to merge the dataset. We can also merge the dataset based on the brand ID/No and then compare the Vendor IDs to ensure that they match along with Vendor Name.</p>

3. <p style = 'color: #1ab9b9ff'> This column seems to again signify a foreign key which allows multiple tables (endinv2017, endinv2018, Purchases, Returns, and Sales) to gather product information from the listings table. While unique to this problem, each row seems to have its own brand id. This can be extremely useful when merging tables in this context as we likely won't use a multi-column to merge the datasets. In addition, once we merge the datasets using Brand ID, we can use it to validate sales price, cost, and product information to other data tables.</p>

4. <p style = 'color: #1ab9b9ff'> This column signifies the primary key in all data tables (except for the Listing table) by "concatenating" the store name and the brand name. Notice that within the dataset, we note that the store identifier is the same as the first 6 digits and the last 4-6 digits is the same as the brand Id within each row. It is then conclusive that each one notes the specific product within each store as a unique identifier (or primary key) within the dataset</p>

## Requirement 2

Inspect the data and answer these specific questions about vendors (HINT: I suggest relying on PriceListing file:)

1. How many unique vendors does BA Toys transact with?
2. How many unique products are there for each vendor?
3. Which vendor’s products produce the largest gross margins? Define the gross margin as sales price less purchase price (cost) divided by sales price.


In [5]:
#  Insert code needed to answer requirement 2 here

# Answering Question 1
print('Number of unique Vendors:', listings['VendorName'].nunique())

#Answering Question 2
print('\nNumber of unique products per Vendor: \n', listings.groupby('VendorName')['ToyDescription'].nunique())

#Answering Question 3
listings['GrossMargin'] = (listings['SalesPrice'] - listings['PurchasePrice']) / listings['SalesPrice']
listings.groupby('VendorName')[['ToyDescription','GrossMargin']].max()

Number of unique Vendors: 4

Number of unique products per Vendor: 
 VendorName
EA              251
FisherPrice    1005
Hasbro         3133
Mattel          463
Name: ToyDescription, dtype: int64


ToyDescription  GrossMargin
VendorName                                                                 
EA                                           WoodWorkers Bench     0.333167
FisherPrice                      Wooden Toys Shivering Penguin     0.708577
Hasbro                              Yokai Watch Plush Jibanyan     0.334171
Mattel       WWE Wrestlemania Heritage Series - Randy Orton...     0.336000

✅ *Insert Requirement 2 Answers:*

1. <p style = 'color: #1ab9b9ff'> 4 companies</p>

2. <p style = 'color: #1ab9b9ff'>EA: 251</br>FisherPrice: 1005 </br>Hasbro: 3133 </br>Mattel: 463</br></p>

3.<p style = 'color: #1ab9b9ff'>EA: WoodWorkers Bench</br>FisherPrice: Wooden Toys Shivering Penguin</br>Hasbro: Yokai Watch Plush Jibanyan</br>Mattel: WWE Wrestlemania Heritage Series - Randy Orton Figure</br></p>


## Requirement 3

Evaluate the external auditors’ claims. Specifically, evaluate sales **by store, by product**.

Focus on these fundamental questions as you answer this question:
1.  Do sales for a given store-product combination ever exceed inventory available for sale? How do you define "Inventory Available for Sales"? (Hint: Beginning Inventory + Purchases + Returns - Ending Inventory; Note that I *add* returns because in the data this is a negative quantity)
2.  If so, does this occur for all products and stores? Or a small number? Describe the stores and products for which you observe sales exceeding available inventory.
3. What is the total dollar amount of the “suspect sales”?
4. What is the average gross margin for the suspect sales?

**HINT**: You'll want to use `pd.merge` to combine datasets, but be careful about merging too early or the dataset will be huge. I suggest using `groupby` to collapse larger datasets (like purchases or sales) down to a more granular level (`InventoryId`).

In [19]:
# Insert code for requirement 3.1 here

# We need to first determine how much product we sold.
# Our first approach will look at how much inventory we sold in our sales datasets vs our inventory dataset.
# After all, we cannot sell products that don't exist. So, if our sales are greater than our inventory, it could indicate fictious 

#Beginning Inventory Aggergation
beginvagg = beginv.groupby(by = ['Store', 'Brand', 'InventoryId'])['OnHand'].sum().reset_index()
#Purchases Inventory Aggregation
purchagg = purch.groupby(by = ['Store', 'Brand', 'InventoryId'])['PurchaseQuantity'].sum().reset_index()
#Returns Aggregation
returnsagg = returns.groupby(by = ['Store', 'Brand', 'InventoryId'])['ReturnedQuantity'].sum().reset_index()
#Ending Inventory Aggregation
endinvagg = endinv.groupby(by = ['Store', 'Brand', 'InventoryId'])['OnHand'].sum().reset_index()

#Merging all the tables as needed
inventory_analysis_table = beginvagg.copy(deep = True)
inventory_analysis_table = inventory_analysis_table.merge(
    purchagg, on =['Store', 'Brand', 'InventoryId'], how = 'outer'
).merge(
    returnsagg, on =['Store', 'Brand', 'InventoryId'], how = 'outer'
).merge(
    endinvagg, on = ['Store', 'Brand', 'InventoryId'], how = 'outer'
).rename(columns = {'OnHand_x': 'StartInv', 'OnHand_y' : 'EndInv'})

# We will fill in the null values in the previous dataset.
# The null values would indicate that no record of that product is there, 
# which could imply no business transactions.
inventory_analysis_table.fillna(0, inplace = True)

inventory_analysis_table['Inventory Available for Sale'] = inventory_analysis_table['StartInv'] + inventory_analysis_table['PurchaseQuantity'] + \
inventory_analysis_table['ReturnedQuantity'] - inventory_analysis_table['EndInv']

#Merging the Sales Table to the Appropriate Aggregation
#Measuring only quantity
sales_analysis_table = sale.groupby(by = ['Store', 'Brand', 'InventoryId'])['SalesQuantity'].sum().reset_index()
comparison_table = inventory_analysis_table.merge(sales_analysis_table, on = ['Store', 'Brand', 'InventoryId'], how = 'outer')

#Filling values with 0 again similar to earlier principles
comparison_table.fillna(0,inplace = True)

#Determing how many products do not match analysis
comparison_table['SalesandInvDiff'] = comparison_table.apply(lambda x: x['SalesQuantity'] - x['Inventory Available for Sale'], axis = 1)

comparison_table[comparison_table['SalesandInvDiff'].abs() > 0]

Store  Brand    InventoryId  StartInv  PurchaseQuantity  \
40684  NJ-4291  89974  NJ-4291-89974       0.0               0.0   
40685  NJ-4291  89975  NJ-4291-89975       0.0               0.0   
40686  NJ-4291  89976  NJ-4291-89976       0.0               0.0   
40687  NJ-4291  89977  NJ-4291-89977       0.0               0.0   
40688  NJ-4291  89978  NJ-4291-89978       0.0               0.0   
40689  NJ-4291  89979  NJ-4291-89979       0.0               0.0   
40690  NJ-4291  89980  NJ-4291-89980       0.0               0.0   
40691  NJ-4291  89981  NJ-4291-89981       0.0               0.0   
40692  NJ-4291  89982  NJ-4291-89982       0.0               0.0   
40693  NJ-4291  89983  NJ-4291-89983       0.0               0.0   
40694  NJ-4291  89984  NJ-4291-89984       0.0               0.0   
40695  NJ-4291  89985  NJ-4291-89985       0.0               0.0   
40696  NJ-4291  89986  NJ-4291-89986       0.0               0.0   
40697  NJ-4291  89987  NJ-4291-89987       0.0               0.0   
40698  NJ-4291  89988  NJ-4291-89988       0.0               0.0   
40699  NJ-4291  89989  NJ-4291-89989       0.0               0.0   
40700  NJ-4291  89990  NJ-4291-89990       0.0               0.0   
40701  NJ-4291  89991  NJ-4291-89991       0.0               0.0   
40702  NJ-4291  89992  NJ-4291-89992       0.0               0.0   
40703  NJ-4291  89993  NJ-4291-89993       0.0               0.0   
40704  NJ-4291  89994  NJ-4291-89994       0.0               0.0   
40705  NJ-4291  89995  NJ-4291-89995       0.0               0.0   
40706  NJ-4291  89996  NJ-4291-89996       0.0               0.0   
40707  NJ-4291  89997  NJ-4291-89997       0.0               0.0   
40708  NJ-4291  89998  NJ-4291-89998       0.0               0.0   
44483  NJ-5264  89974  NJ-5264-89974       0.0               0.0   
44484  NJ-5264  89975  NJ-5264-89975       0.0               0.0   
44485  NJ-5264  89976  NJ-5264-89976       0.0               0.0   
44486  NJ-5264  89977  NJ-5264-89977       0.0               0.0   
44487  NJ-5264  89978  NJ-5264-89978       0.0               0.0   
44488  NJ-5264  89979  NJ-5264-89979       0.0               0.0   
44489  NJ-5264  89980  NJ-5264-89980       0.0               0.0   
44490  NJ-5264  89981  NJ-5264-89981       0.0               0.0   
44491  NJ-5264  89982  NJ-5264-89982       0.0               0.0   
44492  NJ-5264  89983  NJ-5264-89983       0.0               0.0   
44493  NJ-5264  89984  NJ-5264-89984       0.0               0.0   
44494  NJ-5264  89985  NJ-5264-89985       0.0               0.0   
44495  NJ-5264  89986  NJ-5264-89986       0.0               0.0   
44496  NJ-5264  89987  NJ-5264-89987       0.0               0.0   
44497  NJ-5264  89988  NJ-5264-89988       0.0               0.0   
44498  NJ-5264  89989  NJ-5264-89989       0.0               0.0   
44499  NJ-5264  89990  NJ-5264-89990       0.0               0.0   
44500  NJ-5264  89991  NJ-5264-89991       0.0               0.0   
44501  NJ-5264  89992  NJ-5264-89992       0.0               0.0   
44502  NJ-5264  89993  NJ-5264-89993       0.0               0.0   
44503  NJ-5264  89994  NJ-5264-89994       0.0               0.0   
44504  NJ-5264  89995  NJ-5264-89995       0.0               0.0   
44505  NJ-5264  89996  NJ-5264-89996       0.0               0.0   
44506  NJ-5264  89997  NJ-5264-89997       0.0               0.0   
44507  NJ-5264  89998  NJ-5264-89998       0.0               0.0   

       ReturnedQuantity  EndInv  Inventory Available for Sale  SalesQuantity  \
40684               0.0     0.0                           0.0          950.0   
40685               0.0     0.0                           0.0         1430.0   
40686               0.0     0.0                           0.0         1070.0   
40687               0.0     0.0                           0.0         1210.0   
40688               0.0     0.0                           0.0         1200.0   
40689               0.0     0.0                           0.

In [20]:
# Insert code for requirement 3.2 here
potential_issues = comparison_table[comparison_table['SalesandInvDiff'].abs() > 0]
potential_issues = potential_issues.merge(listings, left_on = 'Brand', right_on = 'BrandId')
print("Stores Affected:\n",potential_issues['Store'].unique())
print("Products Affeted:\n", potential_issues['ToyDescription'].unique())

Stores Affected:
 ['NJ-4291' 'NJ-5264']
Products Affeted:
 ['Special Edition: Little People City Skyway (Limited Stock)'
 'Special Edition: Imaginext Power Rangers Blue Ranger & Black Ranger (Limited Stock)'
 'Special Edition: Poppity Pop Turtle (Limited Stock)'
 'Special Edition: Thomas & Friends Wooden Railway Battery-Operated James (Limited Stock)'
 'Special Edition: Thomas & Friends Wooden Railway Series Tidmouth Sheds Deluxe Set (Limited Stock)'
 "Special Edition: Baby's Bandstand Play Gym (Limited Stock)"
 'Special Edition: Thomas & Friends Wooden Railway Battery-Operated Diesel (Limited Stock)'
 'Special Edition: Little People Disney Snow White & the Seven Dwarfs (Limited Stock)'
 'Special Edition: Little People Christmas on Main Street (Limited Stock)'
 'Special Edition: Scoop n Link Bath Boats (Limited Stock)'
 'Special Edition: Little People Lil Movers School Bus (Pink) (Limited Stock)'
 'Special Edition: Little People Soccer Player & Coach (Limited Stock)'
 'Special Edition:

In [21]:
# Insert code for requirement 3.3 here

# Creating a separate document because I don't want to modify earlier code
tempSales = sale.groupby(by = ['Store', 'Brand'])['SalesDollars'].sum()

new_comparison_table = comparison_table[comparison_table['SalesandInvDiff'].abs() > 0].copy(deep = True)
new_comparison_table = new_comparison_table.merge(tempSales, on = ['Store', 'Brand'])
new_comparison_table['SalesDollars'].sum()

2689385.5

In [22]:
# Insert code for requirement 3.4 here

#Getting Brand Names
fraudBrand = new_comparison_table['Brand'].unique()
listings[listings['BrandId'].isin(fraudBrand)]['GrossMargin'].mean()

0.6663020277622759

✅ *Insert Requirement 3 Answers:*

1.<p style = 'color: #1ab9b9ff'> Looking at the quantity amounts, we did notice a few mistakes in the quantity amounts. More specifically, there seems to be two stores affected by fictious sales NJ - 4291 and NJ - 5264. For this problem, we defined Inventory Available for Sale as the Beginning Inventory + Purchases - Inventory - Ending Inventory for the entire year. Looking at a time analysis and a year aggregate did not result in significant differences. Hence, we focused on the year aggregation.
</p>

2.<p style = 'color: #1ab9b9ff'> Our analysis revealed that not many products and stores are affected by this particular over-representation of sales. The overstatement was focused on two stores: NJ - 4291 and NJ - 5264. In addition, the types of products that were most affected by this analysis are the "Special Edition" products, regardless of what type of special product it is
</p>

3.<p style = 'color: #1ab9b9ff'> The total amount would be $2,689,385.50
</p>

4.
<p style = 'color: #1ab9b9ff'> The average profit margin would most likely be round $0.67
</p>

## Requirement 4 (4046 Only )
Conduct a qualitative analysis of fraudulent sales. Specifically, generate and output a CSV file with the likely frauduent sales and review. Identify any red flags you oberve in data.

Note that you should submit your CSV file of fraudulent sales along with your completed jupyter notebook.


> 📌 Please do not zip the files. When files are unzipped after being downloaded from Canvas for grading purposes, your identity information is lost, which may lead to confusion with other students’ submissions.

In [28]:
# Insert code to answer requirement 4

#We have already identified all sales that don't have inventory should be flagged. 
#What does this mean? We can just export the earlier results and get the output that way. 

#Getting Stores Values
flaggedProducts = comparison_table[comparison_table['SalesandInvDiff'].abs() > 0]['InventoryId'].unique()
dfToExport = sale[sale['InventoryId'].isin(flaggedProducts)]
dfToExport.to_csv('Requirement 4.csv', index = False)

✅ *Insert answer for requirement 4*

* <p style = 'color: #1ab9b9ff'> As mentioned earlier, most of the products are special products that appears to have no inventory transactions associated with them. So, there are two stores that mostly affect these transactions (NJ - 4291 and NJ - 5264) and appear to all be sourced by the same vendor (FisherPrice). In addition, all the products are Special Products as well, whcih might have affected normal operations.
</p>


## Requirement 5 (6046 Only)
Conduct a systematic review of the fraudulent sales (i.e., from the original sales records). Specifically, choose **at least 4** of the following questions to analyze, commenting on the difference between fraudulent and legitimate sales. You can use visualizations, tables, or other methods to answer these questions.

Add a code cell for each, as well as a markdown that provides an analysis:

1. Sales quantities
2. Sales prices
3. Sales dates
4. Toy brands
5. Purchase prices
6. Toy descriptions

When done, save your sample of fraudulent sales to a CSV file and submit it along with your completed jupyter notebook.


>  📌 Please do not zip the files. When files are unzipped after being downloaded from Canvas for grading purposes, your identity information is lost, which may lead to confusion with other students’ submissions.

In [ ]:
# Analysis 1


✅ Discussion of Analysis 1:


In [ ]:
# Analysis 2


✅ Discussion of Analysis 2:

In [ ]:
# Analysis 3


✅ Discussion of Analysis 3:

In [ ]:
# Analysis 4


✅ Discussion of Analysis 4:

In [ ]:
timeanalysis.groupby(['Store', 'Brand'])['InventoryChange'].cumsum()

0          5
0          3
1          2
2          0
0          2
          ..
3793413    7
3793414    6
3793415    4
3793416    2
757030     3
Name: InventoryChange, Length: 4605281, dtype: int64

In [ ]:
#I give up, so let's take a look at the specific time periods
from datetime import date
import numpy as np
#There are no dates for the Return table, so we will assume best case scenario
#and that all returns happen at the end of the year
beginv['StartDate'] = date(2017,1,1)
timebeginvagg = beginv.groupby(by = ['Store', 'Brand', 'InventoryId','StartDate'])['OnHand'].sum().reset_index()
timebeginvagg.rename(columns = {'OnHand': 'InventoryChange', 'StartDate': 'Date'}, inplace = True)

purch['ReceivingDate'] = pd.to_datetime(purch['ReceivingDate']).dt.date
timepurchagg = purch.groupby(by = ['Store', 'Brand', 'InventoryId','ReceivingDate'])['PurchaseQuantity'].sum().reset_index()
timepurchagg.rename(columns = {'ReceivingDate' : 'Date', 'PurchaseQuantity': 'InventoryChange'}, inplace = True)

sale['SalesDate'] = pd.to_datetime(sale['SalesDate']).dt.date
timesaleagg = sale.groupby(by = ['Store', 'Brand', 'InventoryId','SalesDate'])['SalesQuantity'].sum().reset_index()
timesaleagg['SalesQuantity'] *= -1
timesaleagg.rename(columns = {'SalesDate': 'Date', 'SalesQuantity': 'InventoryChange'}, inplace = True)

timeanalysis = pd.concat([timebeginvagg, timepurchagg, timesaleagg])
timeanalysis.sort_values(by = ['Store','Brand','Date'], inplace = True)


#Getting the Rolling Sums of each group
timeanalysis['RollingSum'] = timeanalysis.groupby(['Store', 'Brand'])['InventoryChange'].cumsum()
timeanalysis.sort_values(by = 'RollingSum')

Store  Brand    InventoryId        Date  InventoryChange  \
2270531  NJ-5264  89997  NJ-5264-89997  2018-01-31             -300   
2270408  NJ-5264  89986  NJ-5264-89986  2018-01-31             -120   
2270349  NJ-5264  89981  NJ-5264-89981  2018-01-31              -50   
2270407  NJ-5264  89986  NJ-5264-89986  2017-12-31             -220   
2270348  NJ-5264  89981  NJ-5264-89981  2017-12-31             -740   
...          ...    ...            ...         ...              ...   
3105013  PA-1445  62449  PA-1445-62449  2017-11-15             -216   
3159828  PA-1445  75988  PA-1445-75988  2017-11-16             -210   
613517   PA-1445  62449  PA-1445-62449  2017-11-15             2148   
3159827  PA-1445  75988  PA-1445-75988  2017-11-15             -300   
620368   PA-1445  75988  PA-1445-75988  2017-11-15             7478   

         RollingSum  
2270531       -2660  
2270408       -2520  
2270349       -2430  
2270407       -2400  
2270348       -2380  
...             ...  
3105013       14990  
3159828       15024  
613517        15206  
3159827       15234  
620368        15534  

[4605281 rows x 6 columns]

In [ ]:
comparison_table.isnull().sum()

Store                            0
Brand                            0
StartInv                        50
PurchaseQuantity                50
ReturnedQuantity                50
EndInv                          50
Inventory Available for Sale    50
SalesQuantity                    0
SalesandInvDiff                 50
dtype: int64

In [ ]:
comparison_table[(comparison_table['Store'] == 'NJ-5264') & (comparison_table['Brand'] == 89997)]

Store  Brand  StartInv  PurchaseQuantity  ReturnedQuantity  EndInv  \
44506  NJ-5264  89997       NaN               NaN               NaN     NaN   

       Inventory Available for Sale  SalesQuantity  SalesandInvDiff  
44506                           NaN         2660.0              NaN

In [ ]:
timeanalysis['RollingSum'].sort_values().head(5)

2270531   -2660
2270408   -2520
2270349   -2430
2270407   -2400
2270348   -2380
Name: RollingSum, dtype: int64

In [ ]:
timesaleagg

Store  Brand       Date  InventoryChange
0        CT-6253     49 2017-02-02               -2
1        CT-6253     49 2017-02-10               -1
2        CT-6253     49 2017-02-11               -2
3        CT-6253     49 2017-03-11               -1
4        CT-6253     49 2017-03-25               -1
...          ...    ...        ...              ...
3793412  VT-8234  99981 2017-06-10               -1
3793413  VT-8234  99981 2017-11-26               -1
3793414  VT-8234  99981 2017-12-01               -1
3793415  VT-8234  99981 2017-12-03               -2
3793416  VT-8234  99981 2017-12-23               -2

[3793417 rows x 4 columns]

In [ ]:
comparison_table['SalesQuantity'].describe()

count    74838.000000
mean       220.026350
std        753.158741
min          0.000000
25%         20.000000
50%         60.000000
75%        162.000000
max      40070.000000
Name: SalesQuantity, dtype: float64

In [ ]:
sale['InventoryId'].unique().size

74695